### ISSUE TITLE
Unexpected and unwanted caching

### ISSUE THREAD

**Issue by surabujin** [technige/py2neo#683]
I have faced "caching" data in py2neo or in neo4j(I don't have proves to blame only one side of conversation). I don't "see" any pattern in "caching" behavior. Just some random sequence of request lead to fetching stale data.

I don't use ORM extension of py2neo - just 
```python
  with neo_connect.bigin() as tx:
    cursor = tx.tun('...')
    for record in cursor:
      "do some stuff"
```

I not first who faced this problem...

https://stackoverflow.com/questions/49078348/py2neo-and-eventual-consistency
https://stackoverflow.com/questions/34273072/is-py2neo-caching-burning-me

So how can I completely disable all possible node caching in neo4j?

**Comment by keturn** [comment]
One of my application's test cases was just failing when checking the properties of this relationship.

```cypher
MATCH (collection:Collection{uuid:{collection_id}})<-[cc:CircleCollection]-(circle:Circle{uuid:{circle_id}})
RETURN cc
```

When failing, it would return old values for the relationship's properties.
Calling `graph.relationship_cache.clear()` before running the query fixed it.

`Graph.relationship_cache` is not a document attribute.

**Comment by surabujin** [comment]
There is 2 of them in `Graph` class:
```
    node_cache = ThreadLocalEntityCache()
    relationship_cache = ThreadLocalEntityCache()
```

`py2neo devs` if you are trying to mimic sqlalchemy session - you have failed! Now it is extremely unexpectedly and unpredictable.

Why everything about/around neo4j is so sh...t ?..

**Comment by technige** [comment]
The caches are an artifact of the early days of py2neo, when it was a pure HTTP/REST driver and when the HTTP responses were less comprehensive than they are today. Back then, there was much to be gained from saving the extra network hops required to fill in missing information or for retrieving the same information multiple times.

To that end -- and with the release of py2neo v4, which no longer uses the REST API, and which is "Bolt-first" -- it is perfectly reasonable to now question the usefulness of the caches, and the trade-offs that they bring. Indeed, I am planning to look into removing or replacing the caches in a future version, as and when time allows.

Given your comments, though, it's clear you have no desire to engage in constructive conversation. So I'm going to close this thread.

**Comment by surabujin** [comment]
@technige hm... I have too many issues due this "caching" that is not mentioned anywhere (actually first time I faced this issues on your v3... where is was even more hidden than in v4).

You are right I don't want to contribute in anything related to neo4j... I will vote against it in any project. Unfortunately now I had to use it... It is constant pain...

While you are removing caching from your codebase - mention it's presence in documentation and mention how to turn it off (or at least reset it). It can significantly decrease amount of angry people.
